nextspeaker-web/src/components.rs

```diff
@@ -1,30 +1,92 @@
+#[derive(Properties, Debug, PartialEq)]
+pub struct DismissablePanelProps {
+    #[prop_or_default]
+    pub children: Html,
 }
 
 #[styled_component]
-pub fn SimulationPanel(props: &SimulationPanelProps) -> Html {
+pub fn DismissablePanel(props: &DismissablePanelProps) -> Html {
+    let navigator = use_navigator().unwrap();
+    let dismiss = Callback::from(move |_| navigator.push(&Mode::MainView));
     html! {
         <div class={css!("display: flex; background-color: lightgray; flex-direction: column;")}>
             <div class={css!("display: flex; flex-flow: row-reverse;")}>
-                <DismissButton onclick={props.dismiss.clone()} />
+                <DismissButton onclick={dismiss} />
                 <div class={css!("flex: 1;")} />
             </div>
-            <div><h2>{"Simulation of Next Choice"}</h2></div>
-            <div>
-                <button onclick={props.simulate.clone()}>
-                    {format!("run simulation {N_SIM} times")}
-                </button>
-                <SimulationResults results={props.results.clone()} />
-            </div>
+            {props.children.clone()}
+        </div>
+    }
+}
+
+fn ignore_non_candidates(candidates: &Vec<String>, history: &Vec<String>) -> Vec<String> {
+    log!(JsValue::from(&format!("{:?}", candidates)));
+    let candidates: HashSet<_> = candidates.iter().collect();
+    log!(JsValue::from(&format!("{:?}", &history)));
+    let history = history
+        .into_iter()
+        .filter(|h| candidates.contains(h))
+        .cloned()
+        .collect();
+    log!(JsValue::from(&format!("{:?}", &history)));
+    history
+}
+
+#[derive(Properties, PartialEq)]
+pub struct ChooseButtonProps {}
+
+#[styled_component]
+pub fn ChooseButton(_props: &ChooseButtonProps) -> Html {
+    let selected_dispatch = Dispatch::<state::Selected>::new();
+    let candidates = Dispatch::<state::Candidates>::new().get();
+    let (history, history_dispatch) = use_store::<state::History>();
+    let history_halflife = {
+        let hh = Dispatch::<state::HistoryHalflife>::new().get();
+        hh.into_f64()
+    };
+    log!(JsValue::from(&format!(
+        "history_halflife: {}",
+        history_halflife
+    )));
+    let onclick = selected_dispatch.reduce_mut_callback(move |selected| {
+        if !candidates.value.is_empty() {
+            let history = ignore_non_candidates(&candidates.value, &history.value);
+            let new_selection =
+                nextspeaker::choose(&candidates.value, &history, history_halflife).unwrap();
+            history_dispatch.reduce_mut(|h| h.value.push(new_selection.clone()));
+            selected.value = new_selection;
+        }
+    });
+    html! {
+        <button {onclick}>{"choose"}</button>
+    }
+}
+
+#[derive(Properties, PartialEq)]
+
+pub struct SimulationPanelProps {}
+
+#[styled_component]
+pub fn SimulationPanel(_props: &SimulationPanelProps) -> Html {
+    html! {
+        <div>
```

Comment:
Unless you *need* this `<div>`, you can do `<>` to avoid adding an additional element to the dom.